modules/internal/ChapelIteratorSupport.chpl

```diff
@@ -104,6 +104,51 @@ module ChapelIteratorSupport {
       chpl_buildStandInRTT(__primitive("scalar promotion type", t)) );
   }
 
+  // A helper to handle #16027 ex. test/reductions/reduceLoopOfPromoted.chpl
+  pragma "no doc"
+  proc chpl_elemTypeForReducingIterables(x) type {
+
+    // Part 1 - get the first element yielded by 'x'
+    //
+    // An earlier attempt followed the lead of iteratorIndexType()
+    // and called iteratorIndex(_getIterator(x)). Alas in this case
+    // the _iteratorClass returned by _getIterator(x) contains, in turn,
+    // another iterator class. That latter does not get deallocated and leaks.
+    //
+    // Whereas iteratorIndex() invokes ic.advance() and ic.getValue(),
+    // here we go ahead an iterate over this iterator into its first
+    // iteration. Breaking out of the loop with a 'return' somehow causes
+    // proper cleanup of the iterator classes at hand
+    //
+    // This function IS executed at runtime because the returned type
+    // is an array and therefore has a runtime component.
+
+    for i in x {
+      compilerAssert(i.type <= _iteratorRecord); // prevent unintended use
+
+      // The case where the shape is a range can be implemented
+      // analogously to the two versions of chpl__initCopy(_iteratorRecord)
+      // that invoke chpl__initCopy_shapeHelp().
+      if !chpl_iteratorHasDomainShape(i) then
+        compilerError("unsupported elements of the expression being reduced -- they are iterable expressions without a domain shape");
```

Comment:
avoid the too-long line by breaking apart the error with `compilerError("part1 " + "part2 ")` with the two literals on different lines